velox/common/memory/tests/MemoryManagerTest.cpp

```diff
@@ -387,6 +388,122 @@ TEST(MemoryManagerTest, concurrentPoolAccess) {
   pools.clear();
   ASSERT_EQ(manager.numPools(), 0);
 }
+
+TEST(MemoryManagerTest, quotaEnforcement) {
+  struct {
+    int64_t memoryQuotaBytes;
+    int64_t smallAllocationBytes;
+    int64_t largeAllocationBytes;
+    bool expectedMemoryExceedError;
+
+    std::string debugString() const {
+      return fmt::format(
+          "memoryQuotaBytes:{} smallAllocationBytes:{} largeAllocationBytes:{} expectedMemoryExceedError:{}",
+          succinctBytes(memoryQuotaBytes),
+          succinctBytes(smallAllocationBytes),
+          succinctBytes(largeAllocationBytes),
+          expectedMemoryExceedError);
+    }
+  } testSettings[] = {
+      {2 << 20, 1 << 20, 1 << 20, false},
+      {2 << 20, 1 << 20, 2 << 20, true},
+      {2 << 20, 2 << 20, 1 << 20, true},
+      {2 << 20, 3 << 20, 0, true},
+      {2 << 20, 0, 3 << 20, true}};
+
+  for (const auto& testData : testSettings) {
+    SCOPED_TRACE(testData.debugString());
+    std::vector<bool> contiguousAllocations;
+    contiguousAllocations.push_back(true);
+    contiguousAllocations.push_back(false);
+    for (const auto& contiguousAlloc : contiguousAllocations) {
+      SCOPED_TRACE(fmt::format("contiguousAlloc {}", contiguousAlloc));
+      const int alignment = 32;
+      IMemoryManager::Options options;
+      options.alignment = alignment;
+      options.capacity = testData.memoryQuotaBytes;
+      MemoryManager manager{options};
+      auto pool = manager.getPool("quotaEnforcement", MemoryPool::Kind::kLeaf);
+      void* smallBuffer{nullptr};
+      if (testData.smallAllocationBytes != 0) {
+        if ((testData.largeAllocationBytes == 0) &&
+            testData.expectedMemoryExceedError) {
+          VELOX_ASSERT_THROW(pool->allocate(testData.smallAllocationBytes), "");
+          continue;
+        }
+        smallBuffer = pool->allocate(testData.smallAllocationBytes);
+      }
+      if (contiguousAlloc) {
+        ContiguousAllocation contiguousAllocation;
+        if (testData.expectedMemoryExceedError) {
+          VELOX_ASSERT_THROW(
+              pool->allocateContiguous(
+                  AllocationTraits::numPages(testData.largeAllocationBytes),
+                  contiguousAllocation),
+              "");
```

Comment:
The error messages are different for each test case which needs to put it as a test input parameter so it might be a bit complicated. I might just check VELOX exception here :)